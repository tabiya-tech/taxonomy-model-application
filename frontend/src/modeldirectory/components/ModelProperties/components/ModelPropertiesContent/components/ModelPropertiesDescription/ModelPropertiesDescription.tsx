import React from "react";
import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import Box from "@mui/material/Box";
import { useTheme } from "@mui/material";
import TextPropertyField from "src/theme/PropertyFieldLayout/TextPropertyField/TextPropertyField";
import MarkdownPropertyField from "src/theme/PropertyFieldLayout/MarkdownPropertyField/MarkdownPropertyField";

export interface ModelPropertiesDescriptionProps {
  model: ModelInfoTypes.ModelInfo;
}

const uniqueId = "2dc9c69b-bd8f-4b51-bff8-69faa511a316";

export const DATA_TEST_ID = {
  MODEL_PROPERTIES_DESCRIPTION_CONTAINER: `model-properties-description-container-${uniqueId}`,
  MODEL_PROPERTIES_NAME: `model-properties-name-${uniqueId}`,
  MODEL_PROPERTIES_LOCALE: `model-properties-locale-${uniqueId}`,
  MODEL_PROPERTIES_DESCRIPTION: `model-properties-description-${uniqueId}`,
};

export const FIELD_ID = {
  NAME: `name-${uniqueId}`,
  LOCALE: `locale-${uniqueId}`,
  DESCRIPTION: `description-${uniqueId}`,
};

export const FIELD_LABEL_TEXT = {
  LABEL_NAME: "Name",
  LABEL_LOCALE: "Locale",
  LABEL_DESCRIPTION: "Description",
};

/**
 * ModelPropertiesDescription responsible for displaying the basic details of a model
 * like description, locale, etc.
 * @param props
 * @constructor
 */
const ModelPropertiesDescription: React.FC<ModelPropertiesDescriptionProps> = (
  props: Readonly<ModelPropertiesDescriptionProps>
) => {
  const theme = useTheme();

  return (
    <Box
      display="flex"
      flexDirection="column"
      gap={theme.tabiyaSpacing.md}
      data-testid={DATA_TEST_ID.MODEL_PROPERTIES_DESCRIPTION_CONTAINER}
    >
      <TextPropertyField
        label={FIELD_LABEL_TEXT.LABEL_NAME}
        text={`${props.model.name}`}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_NAME}
        fieldId={FIELD_ID.NAME}
      />
      <TextPropertyField
        label={FIELD_LABEL_TEXT.LABEL_LOCALE}
        text={`${props.model.locale.name}(${props.model.locale.shortCode})`}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_LOCALE}
        fieldId={FIELD_ID.LOCALE}
      />
      <MarkdownPropertyField
        label={FIELD_LABEL_TEXT.LABEL_DESCRIPTION}
        text={props.model.description}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_DESCRIPTION}
        fieldId={FIELD_ID.DESCRIPTION}
      />
    </Box>
  );
};

export default ModelPropertiesDescription;
