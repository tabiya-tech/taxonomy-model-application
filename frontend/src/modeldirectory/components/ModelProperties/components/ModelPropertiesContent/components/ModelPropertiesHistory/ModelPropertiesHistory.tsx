import React from "react";
import { useTheme } from "@mui/material";
import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import Box from "@mui/material/Box";
import ModelPropertiesItemDetails from "src/modeldirectory/components/ModelProperties/components/ModelPropertiesItemDetails/ModelPropertiesItemDetails";
import { formattedDate } from "src/utils/userFriendlyDateFormat";

export interface ModelPropertiesHistoryProps {
  model: ModelInfoTypes.ModelInfo;
}

const uniqueId = "08e457ca-63fa-4a92-b5fa-729a444ab827";

export const DATA_TEST_ID = {
  MODEL_PROPERTIES_HISTORY_CONTAINER: `model-properties-history-container-${uniqueId}`,
  MODEL_PROPERTIES_CREATED_DATE: `model-properties-created-date-${uniqueId}`,
  MODEL_PROPERTIES_UPDATED_DATE: `model-properties-updated-date-${uniqueId}`,
};

/**
 * ModelPropertiesHistory responsible for displaying the history of the model
 * like created date, updated date, etc.
 * @param props
 * @constructor
 */

const ModelPropertiesHistory: React.FC<ModelPropertiesHistoryProps> = (
  props: Readonly<ModelPropertiesHistoryProps>
) => {
  const theme = useTheme();

  return (
    <Box
      display="flex"
      flexDirection="column"
      gap={theme.tabiyaSpacing.md}
      data-testid={DATA_TEST_ID.MODEL_PROPERTIES_HISTORY_CONTAINER}
    >
      <ModelPropertiesItemDetails
        title={"Date Created"}
        value={formattedDate(props.model.createdAt)}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_CREATED_DATE}
      />
      <ModelPropertiesItemDetails
        title={"Last Updated"}
        value={formattedDate(props.model.updatedAt)}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_UPDATED_DATE}
      />
    </Box>
  );
};

export default ModelPropertiesHistory;
