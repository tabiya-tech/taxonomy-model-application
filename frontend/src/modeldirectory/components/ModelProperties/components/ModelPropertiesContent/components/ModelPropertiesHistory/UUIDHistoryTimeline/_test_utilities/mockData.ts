import { ModelInfoTypes } from "src//modelInfo/modelInfoTypes";

export function getOneFakeUUIDHistoryDetail(i: number): ModelInfoTypes.ModelHistory {
  return {
    UUID: "e26aca13-aa00-4346-9e83-edaf59314378" + i,
    id: "5e12f537f079fbdcb4e61bf" + i,
    localeShortCode: "ZA",
    name: "Model" + i,
    version: "V1.1." + i,
  };
}

export function getFakeUUIDHistoryDetailsArray(length: number): ModelInfoTypes.ModelHistory[] {
  return Array.from({ length }, (_, i) => getOneFakeUUIDHistoryDetail(i));
}

export function getFakeUUIDHistoryDetailsArrayWithUnresolvedUUIDS(length: number): ModelInfoTypes.ModelHistory[] {
  return Array.from({ length }, (_, i) => {
    const details = getOneFakeUUIDHistoryDetail(i);
    details.id = null;
    details.name = null;
    details.version = null;
    details.localeShortCode = null;
    return details;
  });
}
