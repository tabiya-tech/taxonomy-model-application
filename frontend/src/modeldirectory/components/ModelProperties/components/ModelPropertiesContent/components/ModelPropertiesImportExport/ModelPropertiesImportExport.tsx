import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { useTheme } from "@mui/material";
import React from "react";
import Box from "@mui/material/Box";
import ImportTimeline from "./ImportTimeline/ImportTimeline";
import PropertyFieldLayout from "src/theme/PropertyFieldLayout/PropertyFieldLayout";

export interface ModelPropertiesImportExportProps {
  model: ModelInfoTypes.ModelInfo;
}

const uniqueId = "7804fa94-6151-4a64-8685-cdc3d845063c";

export const DATA_TEST_ID = {
  IMPORT_EXPORT_TAB: `import-export-tab-${uniqueId}`,
};

export const FIELD_ID = {
  IMPORT_FIELD: `import-${uniqueId}`,
};

/**
 * ModelPropertiesImportExport is responsible for showing the import and export process state history of a model in a timeline
 * @param props
 * @constructor
 */

const ModelPropertiesImportExport: React.FC<ModelPropertiesImportExportProps> = (
  props: Readonly<ModelPropertiesImportExportProps>
) => {
  const theme = useTheme();

  return (
    <Box
      display="flex"
      flexDirection="column"
      gap={theme.tabiyaSpacing.md}
      data-testid={DATA_TEST_ID.IMPORT_EXPORT_TAB}
    >
      <PropertyFieldLayout title="Import" fieldId={FIELD_ID.IMPORT_FIELD}>
        <ImportTimeline importProcessState={props.model.importProcessState} />
      </PropertyFieldLayout>
    </Box>
  );
};

export default ModelPropertiesImportExport;
