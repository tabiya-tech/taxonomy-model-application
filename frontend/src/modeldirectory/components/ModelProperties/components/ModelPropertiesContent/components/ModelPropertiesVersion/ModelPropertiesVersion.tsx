import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { useTheme } from "@mui/material";
import Box from "@mui/material/Box";
import React from "react";
import TextPropertyField from "src/theme/PropertyFieldLayout/TextPropertyField/TextPropertyField";

export interface ModelPropertiesVersionProps {
  model: ModelInfoTypes.ModelInfo;
}

const uniqueId = "8d3b0caa-95e2-4f11-a98c-054208e72438";

export const DATA_TEST_ID = {
  MODEL_PROPERTIES_VERSION_CONTAINER: `model-properties-version-container-${uniqueId}`,
  MODEL_PROPERTIES_UUID: `model-properties-uuid-${uniqueId}`,
  MODEL_PROPERTIES_TABIYA_PATH: `model-properties-tabiya-path-${uniqueId}`,
  MODEL_PROPERTIES_PATH: `model-properties-path-${uniqueId}`,
  MODEL_PROPERTIES_VERSION: `model-properties-version-${uniqueId}`,
  MODEL_PROPERTIES_RELEASED: `model-properties-released-${uniqueId}`,
  MODEL_PROPERTIES_RELEASE_NOTES: `model-properties-release-notes-${uniqueId}`,
};

const FIELD_ID = {
  UUID: `uuid-${uniqueId}`,
  TABIYA_PATH: `tabiya-path-${uniqueId}`,
  PATH: `path-${uniqueId}`,
  VERSION: `version-${uniqueId}`,
  RELEASED: `released-${uniqueId}`,
  RELEASE_NOTES: `release-notes-${uniqueId}`,
};
/**
 * ModelPropertiesVersion responsible for displaying the version details of a model
 * like UUID, tabiyaPath, path, version, released, release notes, etc.
 * @param props
 * @constructor
 */

const ModelPropertiesVersion: React.FC<ModelPropertiesVersionProps> = (
  props: Readonly<ModelPropertiesVersionProps>
) => {
  const theme = useTheme();

  return (
    <Box
      display="flex"
      flexDirection="column"
      gap={theme.tabiyaSpacing.md}
      data-testid={DATA_TEST_ID.MODEL_PROPERTIES_VERSION_CONTAINER}
    >
      <TextPropertyField
        label={"UUID"}
        text={`${props.model.UUID}`}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_UUID}
        fieldId={FIELD_ID.UUID}
      />

      <TextPropertyField
        label={"Tabiya Path"}
        text={`${props.model.tabiyaPath}`}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_TABIYA_PATH}
        fieldId={FIELD_ID.TABIYA_PATH}
      />

      <TextPropertyField
        label={"Path"}
        text={`${props.model.path}`}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_PATH}
        fieldId={FIELD_ID.PATH}
      />

      <TextPropertyField
        label={"Version"}
        text={`${props.model.version}`}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_VERSION}
        fieldId={FIELD_ID.VERSION}
      />

      <TextPropertyField
        label={"Released"}
        text={props.model.released ? "true" : "false"}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_RELEASED}
        fieldId={FIELD_ID.RELEASED}
      />

      <TextPropertyField
        label={"Release Notes"}
        text={props.model.releaseNotes}
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_RELEASE_NOTES}
        fieldId={FIELD_ID.RELEASE_NOTES}
      />
    </Box>
  );
};

export default ModelPropertiesVersion;
