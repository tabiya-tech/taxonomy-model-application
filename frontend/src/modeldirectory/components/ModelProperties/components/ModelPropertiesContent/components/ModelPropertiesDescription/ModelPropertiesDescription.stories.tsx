import { getOneFakeModel } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import { Meta, StoryObj } from "@storybook/react";
import ModelPropertiesDescription from "./ModelPropertiesDescription";

const meta: Meta<typeof ModelPropertiesDescription> = {
  title: "ModelDirectory/ModelProperties/ModelPropertiesContentPanels/ModelPropertiesDescription",
  component: ModelPropertiesDescription,
  tags: ["autodocs"],
  args: {
    model: getOneFakeModel(1),
  },
};

export default meta;

type Story = StoryObj<typeof ModelPropertiesDescription>;

export const Shown: Story = {};
