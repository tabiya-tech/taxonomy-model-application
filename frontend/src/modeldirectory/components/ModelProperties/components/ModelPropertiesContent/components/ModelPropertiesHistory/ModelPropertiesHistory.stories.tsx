import { getOneFakeModel } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import { Meta, StoryObj } from "@storybook/react";
import ModelPropertiesHistory from "./ModelPropertiesHistory";

const meta: Meta<typeof ModelPropertiesHistory> = {
  title: "ModelDirectory/ModelProperties/ModelPropertiesContentPanels/ModelPropertiesHistory",
  component: ModelPropertiesHistory,
  tags: ["autodocs"],
  args: {
    model: getOneFakeModel(1),
  },
};

export default meta;

type Story = StoryObj<typeof ModelPropertiesHistory>;

export const Shown: Story = {};
