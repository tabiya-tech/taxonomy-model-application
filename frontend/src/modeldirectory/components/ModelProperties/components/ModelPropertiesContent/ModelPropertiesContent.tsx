import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import TabControl from "src/theme/TabControl/TabControl";
import ModelPropertiesDescription from "./components/ModelPropertiesDescription/ModelPropertiesDescription";
import ModelPropertiesVersion from "./components/ModelPropertiesVersion/ModelPropertiesVersion";
import { Box } from "@mui/material";
import ModelPropertiesHistory from "./components/ModelPropertiesHistory/ModelPropertiesHistory";

export interface ModelPropertiesContentProps {
  model: ModelInfoTypes.ModelInfo;
}

const uniqueId = "ee760560-4d1a-4555-89a7-1c1d0b5cbd5b";

export const DATA_TEST_ID = {
  MODEL_PROPERTIES_CONTENT: `model-properties-content-${uniqueId}`,
  MODEL_PROPERTIES_TABS: `model-properties-tabs-${uniqueId}`,
};

export const MODEL_PROPERTIES_TAB_LABEL = {
  TAB_LABEL_DEFINITION: "Definition",
  TAB_LABEL_VERSION: "Version",
  TAB_LABEL_HISTORY: "History",
};

export const MODEL_PROPERTIES_TAB_ID = {
  TAB_ID_DEFINITION: `model-properties-definition-tab-${uniqueId}`,
  TAB_ID_VERSION: `model-properties-version-tab-${uniqueId}`,
  TAB_ID_HISTORY: `model-properties-history-tab-${uniqueId}`,
};

/**
 * ModelPropertiesContent responsible for creating the tabs (using TabFactory) for the model Properties
 * @param props
 * @constructor
 */
function ModelPropertiesContent(props: Readonly<ModelPropertiesContentProps>) {
  const tabItems = [
    {
      id: MODEL_PROPERTIES_TAB_ID.TAB_ID_DEFINITION,
      label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_DEFINITION,
      panel: <ModelPropertiesDescription model={props.model} />,
    },
    {
      id: MODEL_PROPERTIES_TAB_ID.TAB_ID_VERSION,
      label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_VERSION,
      panel: <ModelPropertiesVersion model={props.model} />,
    },
    {
      id: MODEL_PROPERTIES_TAB_ID.TAB_ID_HISTORY,
      label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_HISTORY,
      panel: <ModelPropertiesHistory model={props.model} />,
    },
  ];

  return (
    <Box data-testid={DATA_TEST_ID.MODEL_PROPERTIES_CONTENT} sx={{ height: "100%" }}>
      <TabControl
        data-testid={DATA_TEST_ID.MODEL_PROPERTIES_TABS}
        items={tabItems}
        aria-label={"Model properties tabs"}
      />
    </Box>
  );
}

export default ModelPropertiesContent;
