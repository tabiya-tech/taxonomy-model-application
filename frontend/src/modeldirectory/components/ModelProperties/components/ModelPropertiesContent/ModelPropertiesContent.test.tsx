// mute the console
import "src/_test_utilities/consoleMock";

import ModelPropertiesContent, {
  DATA_TEST_ID,
  MODEL_PROPERTIES_TAB_ID,
  MODEL_PROPERTIES_TAB_LABEL,
} from "./ModelPropertiesContent";
import { render, screen } from "@testing-library/react";
import { getOneFakeModel } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import ModelPropertiesDescription from "./components/ModelPropertiesDescription/ModelPropertiesDescription";
import TabControl from "src/theme/TabControl/TabControl";
import ModelPropertiesVersion from "./components/ModelPropertiesVersion/ModelPropertiesVersion";
import ModelPropertiesHistory from "./components/ModelPropertiesHistory/ModelPropertiesHistory";
import ModelPropertiesImportExport from "./components/ModelPropertiesImportExport/ModelPropertiesImportExport";

// mock TabControl
jest.mock("src/theme/TabControl/TabControl", () => {
  const actualTabControl = jest.requireActual("src/theme/TabControl/TabControl");
  const actualModelPropertiesContent = jest.requireActual("./ModelPropertiesContent");
  return {
    ...actualTabControl,
    __esModule: true,
    default: jest.fn().mockImplementation(() => {
      return <div data-testid={actualModelPropertiesContent.DATA_TEST_ID.MODEL_PROPERTIES_TABS}> TabControl</div>;
    }),
  };
});

describe("ModelPropertiesContent", () => {
  describe("render tests", () => {
    test("should render correctly with the provided props", async () => {
      // GIVEN a model
      const givenModel = getOneFakeModel(1);

      // WHEN the ModelPropertiesContent is rendered with the given model
      render(<ModelPropertiesContent model={givenModel} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();

      // AND the content to be shown
      const contentElement = screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_CONTENT);
      expect(contentElement).toBeInTheDocument();
      // AND the tabControl to be shown
      const tabControlElement = screen.getByTestId(DATA_TEST_ID.MODEL_PROPERTIES_TABS);
      expect(tabControlElement).toBeInTheDocument();
      // AND the tabControl to have been passed the correct items
      const expectedDescriptionPanel = <ModelPropertiesDescription model={givenModel} />;
      const expectedVersionPanel = <ModelPropertiesVersion model={givenModel} />;
      const expectedHistoryPanel = <ModelPropertiesHistory model={givenModel} />;
      const expectedImportExportPanel = <ModelPropertiesImportExport model={givenModel} />;
      expect(TabControl).toHaveBeenCalledWith(
        {
          items: [
            expect.objectContaining({
              id: MODEL_PROPERTIES_TAB_ID.TAB_ID_DEFINITION,
              label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_DEFINITION,
              panel: expect.objectContaining({
                type: expectedDescriptionPanel.type,
                props: expectedDescriptionPanel.props,
              }),
            }),
            expect.objectContaining({
              id: MODEL_PROPERTIES_TAB_ID.TAB_ID_VERSION,
              label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_VERSION,
              panel: expect.objectContaining({
                type: expectedVersionPanel.type,
                props: expectedVersionPanel.props,
              }),
            }),
            expect.objectContaining({
              id: MODEL_PROPERTIES_TAB_ID.TAB_ID_HISTORY,
              label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_HISTORY,
              panel: expect.objectContaining({
                type: expectedHistoryPanel.type,
                props: expectedHistoryPanel.props,
              }),
            }),
            expect.objectContaining({
              id: MODEL_PROPERTIES_TAB_ID.TAB_ID_IMPORT_EXPORT,
              label: MODEL_PROPERTIES_TAB_LABEL.TAB_LABEL_IMPORT_EXPORT,
              panel: expect.objectContaining({
                type: expectedImportExportPanel.type,
                props: expectedImportExportPanel.props,
              }),
            }),
          ],
          "data-testid": DATA_TEST_ID.MODEL_PROPERTIES_TABS,
          "aria-label": "Model properties tabs",
        },
        {}
      );
      // AND the content should match the snapshot
      expect(contentElement).toMatchSnapshot();
    });
  });
});
