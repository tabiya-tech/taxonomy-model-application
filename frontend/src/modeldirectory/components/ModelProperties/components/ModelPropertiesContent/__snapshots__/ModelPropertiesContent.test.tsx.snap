// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelPropertiesContent render tests should render correctly with the provided props 1`] = `
<div
  class="MuiBox-root css-10klw3m"
  data-testid="model-properties-content-ee760560-4d1a-4555-89a7-1c1d0b5cbd5b"
>
  <div
    data-testid="model-properties-tabs-ee760560-4d1a-4555-89a7-1c1d0b5cbd5b"
  >
     TabControl
  </div>
</div>
`;
